src/mpid/ch4/shm/xpmem/xpmem_impl.h

```diff
@@ -62,40 +125,29 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_XPMEM_seg_regist(int node_rank, size_t size,
         XPMEM_DBG_PRINT("register apid 0x%lx for node_rank %d, segid 0x%lx\n",
                         (uint64_t) segmap->apid, node_rank, (uint64_t) segmap->remote_segid);
     }
-    segmap->refcount++;
-
-    /* Create segment. Both offset and size must be page aligned. */
-    /* TODO: use pre-allocated buffer */
-    MPIR_CHKPMEM_MALLOC(seg, MPIDI_XPMEM_seg_t *, sizeof(MPIDI_XPMEM_seg_t),
-                        mpi_errno, "xpmem seg", MPL_MEM_SHM);
-    seg->offset = MPL_ROUND_DOWN_ALIGN((uint64_t) remote_vaddr,
-                                       (uint64_t) MPIDI_XPMEM_global.sys_page_sz);
-    offset_diff = (off_t) remote_vaddr - seg->offset;
-    seg->size = MPL_ROUND_UP_ALIGN(size + (size_t) offset_diff, MPIDI_XPMEM_global.sys_page_sz);
-    seg->segmap_ptr = segmap;
-
-    xpmem_addr.apid = segmap->apid;
-    xpmem_addr.offset = seg->offset;
-    seg->vaddr = xpmem_attach(xpmem_addr, seg->size, NULL);
-    /* virtual address or failure(-1) */
-    MPIR_ERR_CHKANDJUMP(seg->vaddr == (void *) -1, mpi_errno, MPI_ERR_OTHER, "**xpmem_attach");
-
-    /* TODO: cache registered segments using AV-tree */
+    /* Create segment. Both low and size must be page aligned. */
+    seg_low = MPL_ROUND_DOWN_ALIGN((uint64_t) remote_vaddr,
+                                   (uint64_t) MPIDI_XPMEM_global.sys_page_sz);
+    offset_diff = (off_t) remote_vaddr - seg_low;
+    seg_size = MPL_ROUND_UP_ALIGN(size + (size_t) offset_diff, MPIDI_XPMEM_global.sys_page_sz);
+    seg_high = seg_low + seg_size;
+    MPIDI_XPMEM_seg_avl_tree_search_and_insert_safe(MPIDI_XPMEM_global.treemaps + node_rank,
+                                                    seg_low, seg_high, segmap,
```

Comment:
You need to check the return value of `MPIDI_XPMEM_seg_avl_tree_search_and_insert_safe`.